src/mongo/client/insert_write_operation.cpp

```diff
@@ -66,8 +66,10 @@ namespace mongo {
     }
 
     BSONObj InsertWriteOperation::_ensureId(const BSONObj& doc) {
-        if (doc.hasField("_id"))
+        if (doc.hasField("_id")) {
+            uassert(0, "value of _id element cannot contain any fields starting with $", doc.okForStorage());
```

Comment:
Validating the whole object seems excessive. The point of this is to protect old servers from a certain unchecked failure mode, but only w.r.t. to the _id element.

I think you should re-write this, more like:

```
BSONElement id = doc.getField("_id");
if (!id.eoo()) {
    if (id has substructure) {
        BSONObj idObj = // get element as object
        uassert(0, "...", idObj.okForStorage();
    }
}
```